rcl/src/rcl/wait.c

```diff
@@ -0,0 +1,448 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcl/wait.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "./common.h"
+#include "./stdatomic_helper.h"
+#include "rcl/error_handling.h"
+#include "rcl/time.h"
+#include "rmw/rmw.h"
+
+typedef struct rcl_wait_set_impl_t
+{
+  size_t subscription_index;
+  rmw_subscriptions_t rmw_subscriptions;
+  size_t guard_condition_index;
+  rmw_guard_conditions_t rmw_guard_conditions;
+  size_t timer_index;
+  rcl_allocator_t allocator;
+} rcl_wait_set_impl_t;
+
+rcl_wait_set_t
+rcl_get_zero_initialized_wait_set()
+{
+  static rcl_wait_set_t null_wait_set = {0};
+  return null_wait_set;
+}
+
+static bool
+__wait_set_is_valid(const rcl_wait_set_t * wait_set)
+{
+  return wait_set && wait_set->impl;
+}
+
+static void
+__wait_set_clean_up(rcl_wait_set_t * wait_set, rcl_allocator_t allocator)
+{
+  if (wait_set->subscriptions) {
+    rcl_wait_set_resize_subscriptions(wait_set, 0);
+  }
+  if (wait_set->guard_conditions) {
+    rcl_wait_set_resize_guard_conditions(wait_set, 0);
+  }
+  if (wait_set->timers) {
+    rcl_wait_set_resize_timers(wait_set, 0);
+  }
+  if (wait_set->impl) {
+    allocator.deallocate(wait_set->impl, allocator.state);
+    wait_set->impl = NULL;
+  }
+}
+
+rcl_ret_t
+rcl_wait_set_init(
+  rcl_wait_set_t * wait_set,
+  size_t number_of_subscriptions,
+  size_t number_of_guard_conditions,
+  size_t number_of_timers,
+  rcl_allocator_t allocator)
+{
+  rcl_ret_t fail_ret = RCL_RET_ERROR;
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+  if (__wait_set_is_valid(wait_set)) {
+    RCL_SET_ERROR_MSG("wait_set already initialized, or memory was uninitialized.");
+    return RCL_RET_ALREADY_INIT;
+  }
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    number_of_subscriptions,
+    "number_of_subscriptions cannot be 0", return RCL_RET_INVALID_ARGUMENT);
```

Comment:
I don't remember, I think it used to be that the size of 0 was indicative of not setup, but that doesn't look to be the case anymore. I removed them in 68facdb, but we'll see if this causes an issue once I get to testing this (but it will be the last thing I test since it needs subscriptions which wake up and timers and guard conditions).